### PR TITLE
Route Chat Completions through Anthropic wire compatibility

### PR DESCRIPTION
**Pull request by PierreLeGuen**
## Summary

- add a small, pure `anthropic_compat` crate that converts original OpenAI Chat Completions wire JSON to Anthropic Messages JSON and converts Anthropic responses/streams back
- preserve unknown supported wire fields without a lossy typed request round-trip
- route direct Anthropic Chat Completions through the same raw transport used by the native endpoint
- preserve canonical message content shapes while moving prompt-cache anchors
- reject unsupported Anthropic server tools and one-hour cache TTLs before making a network request
- retain response usage for cache-read and cache-write accounting without allowing zero stream deltas to erase prior counters
- honor `tool_choice: "none"`, omit null descriptions, filter empty content, and keep legacy tool-result image behavior
- emit explicit warnings when unsupported sampling fields are dropped and fall back safely for unknown stop reasons/tool choices
- add differential fixtures for basic system prompts, tools and sampling, cached tool loops, and vision
- add a synthetic live cache-walk script covering Chat Completions, native Messages, `count_tokens`, streaming, and the currently accepted Claude Code beta headers

The converter remains a pure wire-boundary crate. A hot-path allocation benchmark can be added separately if rollout telemetry shows it is needed.

## Validation

- `cargo test --workspace --lib --bins` — passed; `anthropic_compat` has 20 passing tests
- `cargo clippy --workspace --all-targets -- -D warnings`
- `cargo fmt --all -- --check`
- `python3 -m py_compile scripts/test_anthropic_cache_walk.py`
- `git diff --check`

Stack 3 of 3. Depends on https://github.com/nearai/cloud-api/pull/924, which in turn depends on https://github.com/nearai/cloud-api/pull/923.

### CODE REVIEW

**Review by lloydmak99 — APPROVED** [review]
This routes Anthropic Chat Completions through the new pure `anthropic_compat` wire adapter via an `original_request` sidecar. The wiring is sound and privacy-safe: the sidecar is `skip_serializing`/`skip_deserializing`, populated only for direct chat completions (`completions.rs:588`), nulled when auto-redaction rewrites messages (`completions.rs:1495`), and left `None` for the Responses/text-completion paths; conversion failures map to HTTP 400, not 500s. No merge-blocking issues found.

Optional non-blocking follow-ups:
- `crates/anthropic_compat/src/lib.rs:237,247` — empty text now emits `{"type":"text","text":""}` (legacy skipped empties); the `content:"" + tool_calls` / empty-text+image cases could 400 upstream. Rare in practice; consider filtering empty text blocks.
- `crates/anthropic_compat/src/lib.rs:147-171` — OpenAI-namespace `service_tier` values (`auto`/`default`) and a few other params now hard-400 pre-network where they previously passed. Consider downgrading OpenAI-namespace values to warnings and hard-rejecting only Anthropic-specific premium fields.
- `crates/anthropic_compat/src/stream.rs:247` — `update_count` assigns unconditionally where the legacy parser used `max()`, dropping the guard against a `message_delta` zeroing a non-zero cache count from `message_start`. No current regression (standard `message_delta` restates authoritative cumulative usage), but the defensive `max()` is worth keeping.

Checks: reviewed the full diff (26 files) plus surrounding routing/conversion code and prior review threads; `git diff --check` clean; `python3 -m py_compile` on the opt-in live-test script passed. `cargo` build/test/clippy/fmt could not be run (no Rust toolchain in this environment); PR reports `cargo test --workspace --lib --bins`, `cargo clippy -D warnings`, and `cargo fmt --check` passing, with `cargo audit`/`cargo deny` green in CI.

File: crates/anthropic_compat/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1056 @@
...
+
+fn convert_tool_choice(tool_choice: Option<&Value>) -> Result<Option<Value>, CompatError> {
+    let Some(choice) = tool_choice.filter(|value| !value.is_null()) else {
+        return Ok(None);
+    };
+    match choice {
+        Value::String(value) => match value.as_str() {
+            "auto" => Ok(Some(json!({"type": "auto"}))),
+            "required" => Ok(Some(json!({"type": "any"}))),
+            "none" => Ok(None),
```
[the remark this answers is not in the document]
Comment (PierreLeGuen) [reply crates/anthropic_compat/src/lib.rs:679]: Fixed in `03fe6a0d`. `tool_choice: "none"` now bypasses tool conversion entirely and removes the tools array; a regression test also verifies malformed/server tool definitions are ignored when tools are disabled.

[the remark this answers is not in the document]
Comment (PierreLeGuen) [reply crates/anthropic_compat/src/lib.rs:679]: Fixed in `03fe6a0d`. `tool_choice: "none"` removes converted tools so the raw sidecar preserves the service-level no-tools guarantee.

File: crates/anthropic_compat/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+                })?;
+            let input_schema = function
+                .get("parameters")
+                .cloned()
+                .unwrap_or_else(|| json!({"type": "object", "properties": {}}));
+            let mut converted = Map::new();
+            converted.insert("name".to_string(), Value::String(name.to_string()));
+            converted.insert("input_schema".to_string(), input_schema);
+            if let Some(description) = function.get("description").cloned() {
+                converted.insert("description".to_string(), description);
```
[the remark this answers is not in the document]
Comment (PierreLeGuen) [reply crates/anthropic_compat/src/lib.rs]: Fixed in `03fe6a0d`. Explicitly null tool descriptions are omitted from the Anthropic payload, matching the previous typed converter.

File: crates/anthropic_compat/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1056 @@
...
+        .ok_or_else(|| CompatError::at(format!("{path}.type"), "content type is required"))?;
+    let cache_control = object.get("cache_control").cloned();
+
+    match kind {
+        "text" | "input_text" => {
+            let text = object
+                .get("text")
+                .and_then(Value::as_str)
+                .ok_or_else(|| CompatError::at(format!("{path}.text"), "text must be a string"))?;
+            Ok(text_block(text, cache_control))
```
[the remark this answers is not in the document]
Comment (PierreLeGuen) [reply crates/anthropic_compat/src/lib.rs]: Fixed in `03fe6a0d`. Empty text blocks are filtered, and messages that convert to no content are dropped; production-path regression tests cover multimodal and all-empty cases.

File: crates/anthropic_compat/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,1056 @@
...
+        .is_some_and(|tier| tier != "standard_only")
+    {
+        return Err(CompatError::at(
+            "service_tier",
+            "only service_tier=standard_only is supported",
+        ));
+    }
+    for parameter in ["inference_geo", "mcp_servers", "container"] {
+        if request.contains_key(parameter) {
+            return Err(CompatError::at(
```
[the remark this answers is not in the document]
Comment (PierreLeGuen) [reply crates/anthropic_compat/src/lib.rs:197]: Fixed in `03fe6a0d`. Optional premium fields are rejected only when non-null, so explicit null defaults are treated as absent.

**Comment by PierreLeGuen** [comment]
Review follow-up is complete on the current head `03fe6a0d`.

The compatibility path now preserves cumulative stream usage with `max`, falls back safely for unknown stop reasons/tool choices, drops OpenAI-namespace sampling fields with explicit warnings, filters empty text, omits null descriptions, treats null premium fields as absent, and enforces `tool_choice: "none"` by skipping tool conversion entirely. One-hour TTL scanning is allocation-light and restricted to cache-control positions, tool-result images retain legacy behavior, and the old regression coverage now exercises the production converter.

The pure wire-boundary crate has 20 passing tests; the full workspace test suite, workspace clippy, fmt, the live-script syntax check, and diff checks pass. A stream-allocation benchmark remains a separate follow-up if rollout telemetry shows it is needed, keeping this change KISS.

**Review by lloydmak99 — APPROVED** [review]
Routes Chat Completions through the new Anthropic wire-compat converter. The core conversion, streaming, usage/billing, and privacy paths check out, and the earlier round of blocking concerns have all been addressed on the current head.

Verified:
- Usage counters use `max()` to avoid zero-delta clobbering (`stream.rs:243`), and cache reads/writes fold into `prompt_tokens` so they stay billed (`response.rs:129`).
- Unknown `stop_reason` falls back to `"stop"` (`response.rs:179`); `tool_choice:"none"` omits tools and unknown choices map to auto with a warning (`lib.rs:679`).
- `service_tier` is downgraded to a warning; empty text blocks filtered; the one-hour cache scan is allocation-free and position-scoped.
- `original_request` sidecar is `skip_serializing`/`skip_deserializing`, in-memory only, and nulled before dispatch when auto-redact is active (`completions.rs:1498`).

Optional follow-up (non-blocking): a `tool` message with empty or image-only content converts to a `tool_result` with `content: ""` (`lib.rs:335`); worth confirming Anthropic accepts that shape for tools that return no text.

Checks: `git diff --check` clean; Python cache-walk script compiles. Cargo build/test/clippy/fmt were not run (no Rust toolchain in this environment) — PR reports them green in CI.

**Comment by PierreLeGuen** [comment]
Checked the final optional note from Lloyd: empty or image-only `tool` content still becomes `tool_result.content: ""`, exactly matching the legacy typed converter (`extract_content(...).unwrap_or_default()`). I’m keeping that wire-parity behavior in this PR rather than inventing replacement content. The staging cache-walk should include an empty tool-result case before the feature flag is enabled beyond staging; if Anthropic rejects it, that can be handled as a focused compatibility follow-up.

**Review by lloydmak99 — APPROVED** [review]
Routes direct Anthropic Chat Completions through the new pure `anthropic_compat` wire adapter via an in-memory `original_request` sidecar. All previously-raised blocking items are confirmed addressed at the current head, and the privacy path is sound (`original_request` is `skip_serializing`/`skip_deserializing` on both `CompletionRequest` and `ChatCompletionParams`, nulled on auto-redact at `completions.rs:1492-1497`, `None` for Responses/text paths, and only parameter *names* — never content — are debug-logged).

Verified addressed:
- Streaming usage uses `max()` so a zero-delta `message_delta` can't clobber counts (`stream.rs:243`).
- Unknown `stop_reason` falls back to `"stop"` (`response.rs:179`).
- `service_tier` downgraded to a warning rather than hard-reject (`lib.rs:185-194`).
- Regression tests re-pointed at the production converter (`response.rs:263`).

Optional, non-blocking follow-ups:
- `original_request` (raw customer body) is captured for *all* direct chat completions but only consumed by the Anthropic backend — a small memory cost on non-Anthropic paths, not a correctness issue.
- Image-only `tool` messages flatten to `tool_result.content:""` and there are a few extra per-token allocations on the streaming hot path — both already discussed and intentionally deferred.

Checks: `git diff --check` clean; `python3 -m py_compile scripts/test_anthropic_cache_walk.py` passed. Rust build/test/clippy/fmt not run locally (no toolchain present); PR reports them green in CI.

**Review by lloydmak99 — APPROVED** [review]
Routing Chat Completions through the Anthropic wire-compat layer looks sound — the streaming and non-streaming JSON round-trips deserialize cleanly, and the `original_request` privacy sidecar is correctly gated (`#[serde(skip_serializing, skip_deserializing)]`, nulled on auto-redact) and never persisted. No blocking issues; all earlier concerns are confirmed addressed at the current head.

Non-blocking follow-ups (optional):
- Image-only tool results flatten to `content: ""` — acceptable for now but worth revisiting (`lib.rs`).
- Extra per-token allocations on the streaming hot path could be trimmed later (`stream.rs`).
- `original_request` is captured on all direct chat paths, not just Anthropic — minor memory overhead (`completions.rs:1495`).

Checks: `git diff --check` clean; `python3 -m py_compile scripts/test_anthropic_cache_walk.py` passed. `cargo` build/test/clippy/fmt not run locally (no Rust toolchain), but reported green in CI along with `cargo audit`/`cargo deny`.

**Review by lloydmak99 — APPROVED** [review]
This PR routes direct Anthropic Chat Completions through the new pure `anthropic_compat` wire adapter via an in-memory `original_request` sidecar. The conversion preserves the replaced legacy semantics (max_tokens default, temperature/top_p handling, cache-token billing, finish_reason mappings, headers/version/timeout) and every previously-flagged blocking item is addressed at the current head. No blocking correctness, security, or data-loss issue found.

Optional non-blocking follow-ups:
- `original_request` is captured on all direct chat paths but consumed only by the Anthropic backend — a minor per-request memory cost on other backends (`completions.rs:1492-1497`). It is `#[serde(skip_serializing, skip_deserializing)]`, nulled on auto-redact, and `None` on the Responses/text paths, so no data leaves the process.
- Image-only tool results still convert to `content:""` (`lib.rs`) — pre-existing edge case, safe to defer.

Local checks: `git diff --check` clean; `python3 -m py_compile scripts/test_anthropic_cache_walk.py` passed. Cargo test/clippy/fmt were not run locally (no Rust toolchain in this environment); CI reports `cargo test --workspace --lib --bins` (20 `anthropic_compat` unit tests), `cargo clippy -D warnings`, and `cargo fmt --check` green.
